### PR TITLE
feat(macaroon): wire up macaroon root key service

### PR DESCRIPTION
**Pull request by jack-w-shaw**
feat(macaroon): wire up macaroon root key service

We recently implemented a DQLite backed service for macaroon root keys
which implements dbrootkeystore.Backing & dbrootkeystore.ContextBacking.

Use this service to replace the old Mongo-backed backing.

Create an internal macaroon package, transfering core/macaroons
internally, move the ExpirableStorage interface, constructor and
implementation here out of state.
    
We can now drop state entirely from a few places.

Remove the old Mongo implementation

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```
./main.sh -v cmr
```

Cherry pick the following commits to hotfix a known bug with external users and build juju:
- https://github.com/juju/juju/pull/17752/commits/12437cf19617af9808d546e8fc2a950bd95636e3
- https://github.com/juju/juju/pull/17772/commits/1d32c0e144565c6d445d356c226914b818df5f1b

Copy your ec2 credential over to `./cred.yaml` such that:
```
cat ./cred.yaml
credentials:
  aws:
    ec2-robot-external:
      auth-type: access-key
      access-key: ...
      secret-key: ...
```

The run + observe:
```
$ juju bootstrap aws/eu-west-2 aws --config identity-url=https://api.jujucharms.com/identity --config allow-model-access=true
$ juju grant <lp user>@external superuser
$ juju change-user-password
$ juju logout
$ juju login --user <lp user>@external
Welcome, <lp user>@external. You are now logged into "lxd".

Current model set to "admin/controller".

$ juju whoami
Controller:  aws
Model:       admin/controller
User:        <lp user>@external

$ juju add-credential aws -f ./cred.yaml --controller aws --client
Using cloud "aws" from the controller to verify credentials.
Credential "ec2-robot-external" added locally for cloud "aws".

Controller credential "ec2-robot-external" for user "jack-shaw@external" for cloud "aws" on controller "aws" added.
For more information, see ‘juju show-credential aws ec2-robot-external’.

$ juju add-model m --credential ec2-robot-external
Uploading credential 'aws/jack-shaw@external/ec2-robot-external' to controller
Added 'm' model on aws/eu-west-2 with credential 'ec2-robot-external' for user 'jack-shaw'

$ juju deploy ubuntu
$ juju status
Model  Controller  Cloud/Region   Version      Timestamp
m      aws         aws/eu-west-2  4.0-beta4.1  14:46:11+01:00

App     Version  Status   Scale  Charm   Channel        Rev  Exposed  Message
ubuntu           waiting    0/1  ubuntu  latest/stable   24  no       waiting for machine

Unit      Workload  Agent       Machine  Public address  Ports  Message
ubuntu/0  waiting   allocating  0        18.134.209.114         waiting for machine

Machine  State    Address         Inst id              Base          AZ          Message
0        pending  18.134.209.114  i-0790d125c69f15f77  ubuntu@22.04  eu-west-2b  running
```

NOTE: There are still some bugs with external users. They will hopefully be resolved with https://github.com/juju/juju/pull/17752 + other work

### CODE REVIEW

**Comment by hmlanigan** [comment]
@jack-w-shaw moving core/macaroon to internal/macaroon should be a separate PR, including the relocation of ExpirableStorage. A bit confusing to have it all mixed up.

**Comment by jack-w-shaw** [comment]
Yes I agree. A new PR is on the way

**Review by manadart — COMMENTED** [review]
Code looks great, but I had QA issues due to external users as noted.

File: apiserver/stateauthenticator/context_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+// behind these tests is to test the integration of our own macaroon authentication
+// code with the 'third-party' (Canonical controlled) go-macaroon-bakery code.
+//
+// Particularly much of go-macaroon-bakery depends on injecting an implementation
+// of an interface defined within the library (dbrootkeystore.Backing). The ways
+// in which this interface is then used is an implementation detal of the
+// macaroon-bakery.
+//
+// This means that mocks are not fit for purpose, as we end up making assertions
+// against third part implementation details. A stateful fake wis another option,
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply apiserver/stateauthenticator/context_integration_test.go]: Good spot

Fixed

File: apiserver/service.go [hunk 2]
```diff
@@ -30,6 +31,11 @@ type UserService interface {
 	SetPasswordWithActivationKey(ctx context.Context, name string, nonce, box []byte) (userservice.Sealer, error)
 }
 
+type MacaroonService interface {
```
Comment (SimonRichardson) [inline apiserver/service.go:34]: Comment me.

File: apiserver/stateauthenticator/auth.go [hunk 3]
```diff
@@ -55,6 +56,11 @@ type ControllerConfigService interface {
 	ControllerConfig(context.Context) (controller.Config, error)
 }
 
+type MacaroonService interface {
```
Comment (SimonRichardson) [inline apiserver/stateauthenticator/auth.go:59]: Comment me.

File: internal/macaroon/expirablestorage.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+// to expire data added at the specified time.
+type ExpirableStorage interface {
+	bakery.RootKeyStore
+
+	// ExpireAfter returns a new ExpirableStorage that will expire
+	// added items after the specified duration.
+	ExpireAfter(time.Duration) ExpirableStorage
+}
+
+var DefaultExpiration = 24 * time.Hour
```
Comment (SimonRichardson) [inline internal/macaroon/expirablestorage.go:31]: Comment me.

File: internal/macaroon/expirablestorage.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+func (s *storage) Get(ctx context.Context, id []byte) ([]byte, error) {
+	return s.RootKeyStore.Get(ctx, id)
+}
+
+func (s *storage) RootKey(ctx context.Context) ([]byte, []byte, error) {
+	return s.RootKeyStore.RootKey(ctx)
+}
+
+func (s *storage) ExpireAfter(expireAfter time.Duration) ExpirableStorage {
```
Comment (SimonRichardson) [inline internal/macaroon/expirablestorage.go:69]: Comment me.

File: internal/macaroon/expirablestorage.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery/dbrootkeystore"
+	"github.com/juju/errors"
+)
+
+// Functions defined as variables so they can be overridden
+// for testing.
+var (
+	Clock dbrootkeystore.Clock
+)
```
Comment (SimonRichardson) [inline internal/macaroon/expirablestorage.go:19]: Pass the Clock a as an optional dependency to the `NewExpirableStorage` . That way we don't need to patch the function out during tests.

File: internal/worker/httpserverargs/worker.go [hunk 7]
```diff
@@ -184,15 +186,47 @@ func (b *managedServices) UpdateLastModelLogin(ctx context.Context, name string,
...
+	return b.macaroonService.GetKeyContext(ctx, id)
+}
+
+func (b *managedServices) FindLatestKeyContext(ctx context.Context, createdAfter, expiresAfter, expiresBefore time.Time) (dbrootkeystore.RootKey, error) {
+	return b.macaroonService.FindLatestKeyContext(ctx, createdAfter, expiresAfter, expiresBefore)
+}
+
+func (b *managedServices) InsertKeyContext(ctx context.Context, key dbrootkeystore.RootKey) error {
+	return b.macaroonService.InsertKeyContext(ctx, key)
+}
```
Comment (SimonRichardson) [inline internal/worker/httpserverargs/worker.go:210]: Comment on all of the non-commented methods.

File: internal/macaroon/expirablestorage.go (hunk 6 above)
Comment (jack-w-shaw) [reply internal/macaroon/expirablestorage.go]: Yeah agree.

However, making a clock mandatory turned out to be by far the cleanest way to do this.

**Review by manadart — APPROVED** [review]
All good here. CMRs work too.

Nice job.

**Pull request merged**